### ISSUE TITLE
Installation issue: hpctoolkit

### ISSUE THREAD

**Issue by eschnett** [LLNL/spack#23667]
### Steps to reproduce the issue

```console
$ spack install hpctoolkit %gcc @11.1.0 target=skylake_avx512
```

I receive errors of the kind
```
     2175    /home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/gcc-11.1.0-vvn67u5lrxu5r6ygab
             l722234lfcdix7/include/c++/11.1.0/tr1/riemann_zeta.tcc: In function '_Tp std::__detail::__hurwitz_zeta_glob(_Tp, _Tp)'
             :
  >> 2176    /tmp/eschnetter/spack-stage/spack-stage-hpctoolkit-2021.03.01-ugmvtghcrx5xkf4764yw4xj6igxubt2j/spack-src/src/lib/prof/
             Metric-AExpr.hpp:100:18: error: expected unqualified-id before '(' token
     2177      100 | #define epsilon  (0.000001)
```

Building with `gcc @10.3.0 target=skylake_avx512` works fine.

### Information on your system

```
$ spack debug report
* **Spack:** 0.16.1-2651-fc392d2f56
* **Python:** 3.6.9
* **Platform:** linux-ubuntu18.04-skylake_avx512
* **Concretizer:** original
```

### Additional information

* [spack-build-out.txt](https://github.com/spack/spack/files/6489735/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/spack/spack/files/6489736/spack-build-env.txt)

@mwkrentel

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by mwkrentel** [comment]
@eschnett This kind of error usually is the result of a newer gcc/g++ becoming
more strict on some type of usage.  However, in this case, the error is,
"expected unqualified-id before '(' token" and the offending line is:

```
#define epsilon  (0.000001)
```

I guess I don't see what the problem is here.

Anyway, I'll compile up 11.x and see if I get the same error.

Thanks for reporting this.

**Comment by mwkrentel** [comment]
I can confirm the build error with gcc 11.x.  The C++ header file
<limits> uses the name 'epsilon' in some templates which I think is
the real conflict, although the error message doesn't report that.

It's easy enough to fix (change epsilon to a different name).  But
fixing this exposed several other C++ usages that gcc 11.x didn't
like.  So, it may take a few days to clean them out with 11.x.

We were planning on a new release in the next few days and we thought
we were just about ready.  But now we'll hold the release to fix these
issues.

**Comment by mwkrentel** [comment]
@eschnett I fixed the build issues with gcc 11.x.  If you'd like to try it out,
you should be able to install `hpctoolkit@master %gcc@11.1.0`.

We're planning on a new release in a few days.  I'll update the
hpctoolkit recipe at that time.

**Comment by eschnett** [comment]
I confirm that the master branch builds with GCC 11.

**Comment by mwkrentel** [comment]
I have submitted PR #24017  which fixes this for rev 2021.05.15 and
backports a patch for earlier revs.
